Accept aliases in exception patterns

xclerc opened the pull request:
It is currently not possible to use aliases in exception patterns; for instance
the following code:

    match f 0 with
    | y -> print_int y
    | exception Failure _ as f -> raise f

is refused with the following message:

    Error: Exception patterns must be at the top level of a match case

while the exception pattern is seemingly top level (the alias making it appear
below top level).

xclerc committed:
Accept aliases in exception patterns.

yallop reviewed (changes requested):
@xclerc, could you please add a test case?

shindere commented:
yallop (2017/08/10 05:50 -0700):
> yallop requested changes on this pull request.
> 
> @xclerc, could you please add a test case?

How about a Changes entry? And perhaps a documentation update?

xclerc committed:
Update for GPR#1283.

xclerc committed:
Test for GPR#1283.

xclerc commented:
I have added a test and updated the Changes file (sorry for failing
to do so in the first place).

Regarding the documentation, I have mixed feelings: one may argue
that `exception (E _ as e)` is actually top level (the fact that it is not in
the compiler being merely an artifact of the parsed tree), and hence
that the current documentation is right. What is your take on it?

shindere commented:
Xavier Clerc (2017/08/10 15:40 +0000):
> I have added a test and updated the Changes file

Thanks a lot!

> (sorry for failing to do so in the first place).

No problem! Sometimes it's also good to wait and see whether the
principle of the PR is fine to other devs before doing these tasks to
avoid doing them for nothing, IMO.

> Regarding the documentation, I have mixed feelings: one may argue
> that `exception (E _ as e)` is actually top level (the fact that it is not in
> the compiler being merely an artifact of the parsed tree), and hence
> that the current documentation is right. What is your take on it?

Frankly speaking, I don't know. It's probably better if language / doc
experts can give their opinions.

yallop commented:
Some clarification on the description: it is currently possible to aliases the exceptions matched in exception patterns, but parentheses are needed.  The example in the description can be written as follows with existing OCaml releases:

```ocaml
match f 0 with
| y -> print_int y
| exception (Failure _ as f) -> raise f
```

However, the current relative precedences of `exception` and `as` mean that the `exception` clause cannot be written without parentheses.  So this is a small change to the syntax rather than an entirely new capability.  Besides compatibility issues -- code written with this syntax won't work on old versions of OCaml -- I think the change is essentially harmless (since the syntax is currently disallowed, and since there doesn't seem to be a useful meaning for the other possible interpretation, namely aliasing the full exception pattern), but I don't see any strong arguments in favour of making the change, either.

In terms of the implementation, I think it would be better to change the precedence in the parser, if possible, rather than changing `split_cases` in the typechecker.

yallop commented on testsuite/tests/match-exception/aliases.ml:
```diff
@@ -0,0 +1,23 @@
+  let print_nat_pred n =
+    match nat_pred n with
+    | m -> print_int m
+    | exception (Failure _ as f) -> raise f
```
This case should drop the parentheses to test the new functionality.

xclerc replied:
(test updated)

xclerc committed:
Update test to fail on earlier versions.

xclerc commented:
You are definitely right. I completely missed the ability to use aliases
by just adding parentheses (how embarrassing). Please feel free to
close the PR.
